redwood-compose/src/commonMain/kotlin/app/cash/redwood/compose/WidgetApplier.kt

```diff
@@ -59,37 +64,102 @@ internal class WidgetApplier<W : Any>(
   }
 
   override fun remove(index: Int, count: Int) {
     check(!closed)
 
-    val current = current as ChildrenWidget
-    current.children!!.remove(index, count)
+    val current = current as ChildrenNode
+    current.children.remove(index, count)
   }
 
   override fun move(from: Int, to: Int, count: Int) {
     check(!closed)
 
-    val current = current as ChildrenWidget
-    current.children!!.move(from, to, count)
+    val current = current as ChildrenNode
+    current.children.move(from, to, count)
   }
 
   override fun onClear() {
     closed = true
   }
 }
+
+/** @suppress For generated code usage only. */
+@RedwoodCodegenApi
+public sealed interface Node<W : Any>
+
+/**
+ * A node which wraps a [Widget] and also holds onto the [Widget.Children] in which the widget
+ * is placed. The generics of this class are a little funky in order to avoid casts of the widget
+ * in the generated composables.
+ *
+ * @suppress For generated code usage only.
+ */
+@RedwoodCodegenApi
+public class WidgetNode<W : Widget<V>, V : Any>(
+  public val widget: W,
+) : Node<W> {
+  public var container: Widget.Children<V>? = null
+
+  public companion object {
+    public val SetLayoutModifiers: WidgetNode<*, *>.(LayoutModifier) -> Unit = {
+      widget.layoutModifiers = it
+      container?.onLayoutModifierUpdated()
+    }
+  }
+}
+
+/**
+ * A synthetic widget which allows the applier to differentiate between multiple groups of children.
+ *
+ * Compose's tree assumes each node only has single list of children. Or, put another way, even if
+ * you apply multiple children Compose treats them as a single list of child nodes. In order to
+ * differentiate between these children lists we introduce synthetic nodes. Every real node which
+ * supports one or more groups of children will have one or more of these synthetic nodes as its
+ * direct descendants. The nodes which are produced by each group of children will then become the
+ * descendants of those synthetic nodes.
+ *
+ * This node has two valid states:
+ * 1. Non-null accessor and null children. This is the state when created but not inserted in
+ *    the node tree.
+ * 2. Null accessor and non-null children. Once inserted into the tree, the accessor is used to
+ *    fetch the appropriate children reference from the parent widget.
+ *
+ * Transition from 1 to 2 by calling [attachTo]. This may only be done once, and you
+ * cannot transition back to 1 afterwards.
+ */
+@RedwoodCodegenApi
+internal class ChildrenNode<W : Any> private constructor(
+  private var _accessor: ((Widget<W>) -> Widget.Children<W>)?,
+  private var _children: Widget.Children<W>?,
+) : Node<W> {
+  constructor(accessor: (Widget<W>) -> Widget.Children<W>) : this(accessor, null)
+  constructor(children: Widget.Children<W>) : this(null, children)
+
+  val children: Widget.Children<W> get() = _children!!
+
+  fun attachTo(parent: Widget<W>) {
+    _children = checkNotNull(_accessor).invoke(parent)
+    _accessor = null
```

Comment:
Does accessor need `_` ?

Reply:
Nope! Forgot I added the function to attach rather than doing it at the call site.